src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -699,11 +660,50 @@ private static int NextSiblingIndex(in RenderTreeFrame frame, int frameIndex)
                 {
                     if (oldFrame.ComponentTypeField == newFrame.ComponentTypeField)
                     {
-                        UpdateRetainedChildComponent(
-                            ref diffContext,
-                            oldFrameIndex,
-                            newFrameIndex);
-                        diffContext.SiblingIndex++;
+                        // As an important rendering optimization, we want to skip parameter update
+                        // notifications if we know for sure they haven't changed/mutated. The
+                        // "MayHaveChangedSince" logic is conservative, in that it returns true if
+                        // any parameter is of a type we don't know is immutable. In this case
+                        // we call SetParameters and it's up to the recipient to implement
+                        // whatever change-detection logic they want. Currently we only supply the new
+                        // set of parameters and assume the recipient has enough info to do whatever
+                        // comparisons it wants with the old values. Later we could choose to pass the
+                        // old parameter values if we wanted. By default, components always rerender
+                        // after any SetParameters call, which is safe but now always optimal for perf.
+
+                        // When performing hot reload, we want to force all components to re-render.
+                        // We do this using two mechanisms - we call SetParametersAsync even if the parameters
+                        // are unchanged and we ignore ComponentBase.ShouldRender.
+                        // Furthermore, when a hot reload edit removes component parameters, the component should be
+                        // disposed and reinstantiated. This allows the component's construction logic to correctly
+                        // re-initialize the removed parameter properties.
+
+                        var oldParameters = new ParameterView(ParameterViewLifetime.Unbound, oldTree, oldFrameIndex);
+                        var newParametersLifetime = new ParameterViewLifetime(diffContext.BatchBuilder);
+                        var newParameters = new ParameterView(newParametersLifetime, newTree, newFrameIndex);
+                        var isHotReload = HotReloadManager.Default.MetadataUpdateSupported && diffContext.Renderer.IsRenderingOnMetadataUpdate;
```

Comment:
I wonder if it would make sense to add `IsHotReload` as a flag on `DiffContext` so it can be evaluated just once at the start of each diff process, instead of once for each component we're retaining.

Reply:
Is that going to have a meaningful performance impact? (this is stripped away in production isn't it?)

Reply:
Oh yeah, good point.

In fact that might be a reason why *not* making a separate flag is better, i.e., depending on how smart the trimmer is about static analysis. For example, whether:

```cs
if (someValueThatTheTrimmerWillEvaluateAsFalse)
{
   // ...
}
```

and 

```cs
var shouldDoStuff = someValueThatTheTrimmerWillEvaluateAsFalse;
if (shouldDoStuff)
{
   // ...
}
```

... are actually equivalent, or whether the first is better because the entire `if(false)` is eliminated. I see that the earlier version of this code by Pranav did inline the trimmer-replacable value into the `if` test. Further, I don't know how smart the interpreter is about detecting and eliding `if(false)` blocks during its MSIL->interp-instructions pass.

In this case it's probably not something to worry about, just an interesting thought.